src/resolution/resolver.ts

```diff
@@ -51,103 +31,86 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
-        const isRequestSingleton = binding.scope === BindingScopeEnum.Request;
 
-        if (isSingleton && binding.activated) {
-            return binding.cache;
-        }
+        return _resolveBinding<T>(requestScope, request, binding);
+    }
+};
 
-        if (
-            isRequestSingleton &&
-            requestScope !== null &&
-            requestScope.has(binding.id)
-        ) {
-            return requestScope.get(binding.id);
-        }
+const _resolveFactoryFromBinding = <T>(
+    binding:interfaces.Binding<T>,
+    context:interfaces.Context
+): T | Promise<T> => {
+    const factoryDetails = _getFactoryDetails(binding) as FactoryDetails;
+    return _tryAndThrowErrorIfStackOverflow(
+        () => (factoryDetails.factory as FactoryTypeFunction).bind(binding)(context),
+        () => new Error(
+        ERROR_MSGS.CIRCULAR_DEPENDENCY_IN_FACTORY(factoryDetails.factoryType, context.currentRequest.serviceIdentifier.toString())
+    ));
+}
 
-        if (binding.type === BindingTypeEnum.ConstantValue) {
-            result = binding.cache;
-            binding.activated = true;
-        } else if (binding.type === BindingTypeEnum.Function) {
-            result = binding.cache;
+const _getResolvedFromBinding = <T>(
+    requestScope: interfaces.RequestScope,
+    request: interfaces.Request,
+    binding:interfaces.Binding<T>): T | Promise<T> => {
+    let result: T | Promise<T> | undefined;
+    const childRequests = request.childRequests;
+    _ensureFullyBound(binding);
+    switch(binding.type){
+        case "ConstantValue":
+        case "Function":
+            result = binding.cache as T | Promise<T>;
             binding.activated = true;
-        } else if (binding.type === BindingTypeEnum.Constructor) {
-            result = binding.implementationType as unknown as T;
-        } else if (binding.type === BindingTypeEnum.DynamicValue && binding.dynamicValue !== null) {
-            result = invokeFactory(
-                "toDynamicValue",
-                binding.serviceIdentifier,
-                () => (binding.dynamicValue as (context: interfaces.Context) => any)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Factory && binding.factory !== null) {
-            result = invokeFactory(
-                "toFactory",
-                binding.serviceIdentifier,
-                () => (binding.factory as interfaces.FactoryCreator<any>)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Provider && binding.provider !== null) {
-            result = invokeFactory(
-                "toProvider",
-                binding.serviceIdentifier,
-                () => (binding.provider as interfaces.Provider<any>)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Instance && binding.implementationType !== null) {
-            result = resolveInstance(
+            break;
+        case "Constructor":
```

Comment:
We prefer `BindingTypeEnum`. I know it's not a real enum, we will iterate though this

Reply:
Ok